### PR TITLE
discard optimization

### PR DESCRIPTION
**Pull request by yjaelex**
This optimization will check blocks with a single discard call and
move the discrad call outside and remove the block.

### CODE REVIEW

File: patch/llpcPatchPeepholeOpt.cpp [hunk 1]
```diff
@@ -30,16 +30,31 @@
...
 
 using namespace Llpc;
 using namespace llvm;
 
+namespace llvm
+{
+
+namespace cl
+{
+    // -enable-discard-opt: Enable the optimization for "kill" intrinsic.
```
Comment (amdrexu) [inline patch/llpcPatchPeepholeOpt.cpp]: 4 space indentation is needless. Please reference other options.

File: patch/llpcPatch.h [hunk 2]
```diff
@@ -100,7 +100,7 @@ llvm::ModulePass* CreatePatchInOutImportExport();
 llvm::ModulePass* CreatePatchLlvmIrInclusion();
 llvm::FunctionPass* CreatePatchLoopUnrollInfoRectify();
 llvm::ModulePass* CreatePatchNullFragShader();
-llvm::FunctionPass* CreatePatchPeepholeOpt();
+llvm::FunctionPass* CreatePatchPeepholeOpt(bool enKillOpt = false);
```
Comment (amdrexu) [inline patch/llpcPatch.h]: enKillOpt  -> enableDiscardOpt

File: patch/llpcPatchPeepholeOpt.h [hunk 3]
```diff
@@ -69,7 +69,14 @@ class PatchPeepholeOpt final:
     public llvm::InstVisitor<PatchPeepholeOpt>
 {
 public:
-    explicit PatchPeepholeOpt();
+    explicit PatchPeepholeOpt(bool enKillOpt);
```
Comment (amdrexu) [inline patch/llpcPatchPeepholeOpt.h]: enKillOpt -> enableDiscardOpt

File: patch/llpcPatchPeepholeOpt.h [hunk 4]
```diff
@@ -69,7 +69,14 @@ class PatchPeepholeOpt final:
...
+
+    PatchPeepholeOpt()
+        :
+        FunctionPass(ID)
+    {
+
```
Comment (amdrexu) [inline patch/llpcPatchPeepholeOpt.h]: Remove this line

File: patch/llpcPatchPeepholeOpt.h [hunk 5]
```diff
@@ -93,6 +101,7 @@ class PatchPeepholeOpt final:
     // -----------------------------------------------------------------------------------------------------------------
 
     llvm::SmallVector<llvm::Instruction*, 8> m_instsToErase;
+    bool m_enKillOpt;
```
Comment (amdrexu) [inline patch/llpcPatchPeepholeOpt.h]: m_enKillOpt -> m_enableDiscardOpt; // Whether to enable the optimization for "kill" intrinsic

Add the missing comments

File: patch/llpcPatchPeepholeOpt.cpp [hunk 6]
```diff
@@ -846,6 +864,73 @@ void PatchPeepholeOpt::visitPHINode(
...
+        {
+            auto pTerminator = pPredBlock->getTerminator();
+            BranchInst* pBranch = dyn_cast<BranchInst>(pTerminator);
+            if (pBranch && pBranch->isConditional())
+            {
+                auto pCond = pBranch->getCondition();
+                auto pTrueBlock = dyn_cast<BasicBlock>(pBranch->getSuccessor(0));
+                auto pFalseBlock = dyn_cast<BasicBlock>(pBranch->getSuccessor(1));
+                auto pNewKill = dyn_cast<CallInst>(callInst.clone());
+                Context* pContext = static_cast<Context*>(&callInst.getContext());
```
Comment (amdrexu) [inline patch/llpcPatchPeepholeOpt.cpp]: Add a blank line after this line

**Comment by amdvlk-admin** [comment]
### Test summary for commit 56f2352f38f817f72d9ffd7ea7a705ff33c1dc90
<details open><summary>Driver commits used in build</summary><ul>

<li>PAL: dev 6c8eaa257e6216437fdfe3f17d418eccfe42e0bd</li>
<li>SPVGEN: dev e422cc7f3202b2443c62571349ad703f59027c72</li>
<li>XGL: dev eee58c8e482ac4a6fdc40452cb4ad744395d0f74</li>
<li>LLVM: amd-vulkan-dev 8118692e8aefbd95f791ff49beddb8ce90e8927c</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 900/900 (100.0%)</li>
<li>Failed: 0/900 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 366312/366312 (100.0%)</li>
<li>Failed: 0/366312 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 366312/366312 (100.0%)</li>
<li>Failed: 0/366312 (0.0%)</li>
</ul></details>


</ul></details>

File: patch/llpcPatchPeepholeOpt.h [hunk 7]
```diff
@@ -93,6 +100,7 @@ class PatchPeepholeOpt final:
     // -----------------------------------------------------------------------------------------------------------------
 
     llvm::SmallVector<llvm::Instruction*, 8> m_instsToErase;
+    bool m_enableDiscardOpt;
```
Comment (amdrexu) [inline patch/llpcPatchPeepholeOpt.h]: Add the comments: // Whether to enable the optimization for "kill" intrinsic

**Comment by amdvlk-admin** [comment]
### Test summary for commit 1447a8215d2a30a79e844d06d913501a4ca69a96
<details open><summary>Driver commits used in build</summary><ul>

<li>PAL: dev 6c8eaa257e6216437fdfe3f17d418eccfe42e0bd</li>
<li>SPVGEN: dev e422cc7f3202b2443c62571349ad703f59027c72</li>
<li>XGL: dev eee58c8e482ac4a6fdc40452cb4ad744395d0f74</li>
<li>LLVM: amd-vulkan-dev 8118692e8aefbd95f791ff49beddb8ce90e8927c</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 900/900 (100.0%)</li>
<li>Failed: 0/900 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 366312/366312 (100.0%)</li>
<li>Failed: 0/366312 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 366312/366312 (100.0%)</li>
<li>Failed: 0/366312 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit ac60dfaf95566cfc488aced4ec40331adfc7d0bd
<details open><summary>Driver commits used in build</summary><ul>

<li>PAL: dev 6c8eaa257e6216437fdfe3f17d418eccfe42e0bd</li>
<li>SPVGEN: dev e422cc7f3202b2443c62571349ad703f59027c72</li>
<li>XGL: dev eee58c8e482ac4a6fdc40452cb4ad744395d0f74</li>
<li>LLVM: amd-vulkan-dev 8118692e8aefbd95f791ff49beddb8ce90e8927c</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 900/900 (100.0%)</li>
<li>Failed: 0/900 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 366312/366312 (100.0%)</li>
<li>Failed: 0/366312 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 366312/366312 (100.0%)</li>
<li>Failed: 0/366312 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit fcabcdbfcbea0a7886d56cc8693d494883153987
<details open><summary>Driver commits used in build</summary><ul>

<li>PAL: dev 6c8eaa257e6216437fdfe3f17d418eccfe42e0bd</li>
<li>SPVGEN: dev e422cc7f3202b2443c62571349ad703f59027c72</li>
<li>XGL: dev eee58c8e482ac4a6fdc40452cb4ad744395d0f74</li>
<li>LLVM: amd-vulkan-dev 8118692e8aefbd95f791ff49beddb8ce90e8927c</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 900/900 (100.0%)</li>
<li>Failed: 0/900 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 366312/366312 (100.0%)</li>
<li>Failed: 0/366312 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 366312/366312 (100.0%)</li>
<li>Failed: 0/366312 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**